File: 27-04-2024/Merge Sort on Doubly Linked List.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Node {
    int data;
    Node next, prev;

    Node(int key) {
        data = key;
        next = prev = null;
    }
}

class Driverclass {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        while (t-- > 0) {
            int n = sc.nextInt();
            int a1 = sc.nextInt();
            Node head = new Node(a1);
            Node temp = head;

            for (int i = 1; i < n; i++) {
                int a = sc.nextInt();
                Node n1 = new Node(a);
                n1.prev = temp;
                temp.next = n1;
                temp = n1;
            }

            head = new Solution().sortDoubly(head);
            printList(head);
        }
    }

    public static void printList(Node node) {
        Node temp = node;
        while (node != null) {
            System.out.print(node.data + " ");
            temp = node;
            node = node.next;
        }
        System.out.println();
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.prev;
        }
        System.out.println();
    }
}

// } Driver Code Ends


// User function Template for Java

/*
class Node
{
    int data;
    Node next, prev;
    Node(int data)
    {
        this.data = data;
        next = prev = null;
    }
}
*/
class Solution {
    // Function to sort the given doubly linked list using Merge Sort.
    static Node sortDoubly(Node head) {
      //Your Code Start
         if(head == null || head.next == null)
            return head;
            
        Node slow = head;
        Node fast = head;
        
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        
        Node mid = slow;
        Node beforeMid = slow.prev;
        beforeMid.next = null;
        mid.prev = null;
        
        Node left = sortDoubly(head);
        Node right = sortDoubly(mid);
        
        return mergeDoubly(left,right);
    }
    
    static Node mergeDoubly(Node a,Node b){
        if(a == null)
            return b;
        if(b == null)
            return a;
            
        Node head = null;
        Node tail = null;
        Node temp = null;
        
        if(a.data <= b.data){
            head = a;
            tail = a;
            a = a.next;
        }
        else{
            head = b;
            tail = b;
            b = b.next;
        }
        
        while(a != null && b != null){
            if(a.data <= b.data){
                tail.next = a;
                temp = tail;
                tail = a;
                tail.prev = temp;
                a = a.next;
            }
            else{
                tail.next = b;
                temp = tail;
                tail = b;
                tail.prev = temp;
                b = b.next;
            }
        }
        
        if(a == null){
             tail.next = b;
             temp = tail;
             tail = b;
             tail.prev = temp;
        }
        if(b == null){
             tail.next = a;
             temp = tail;
             tail = a;
             tail.prev = temp;
        }
            
        return head;
 //Your Code end
    }

}

File: 27-04-2024/Merge Sort on Doubly Linked List.py
#User function Template for python3

'''
class Node:
	def __init__(self, data):
		self.data = data
		self.next = None
		self.prev = None
'''
#Your Code Start
class Solution():
#Function to sort the given doubly linked list using Merge Sort.
 
    def sortDoubly(self, head):
        def mergesort(sta):
            if not sta or not sta.next:
                return sta
            
            # Function to find the midpoint of the list
            def find_mid(start):
                slow = start
                fast = start.next
                while fast and fast.next:
                    slow = slow.next
                    fast = fast.next.next
                return slow
            
            # Function to merge two sorted lists
            def merge(left, right):
                dummy = Node(None)
                cur = dummy
                while left and right:
                    if left.data <= right.data:
                        cur.next = left
                        left.prev = cur
                        left = left.next
                    else:
                        cur.next = right
                        right.prev = cur
                        right = right.next
                    cur = cur.next
                if left:
                    cur.next = left
                    left.prev = cur
                if right:
                    cur.next = right
                    right.prev = cur
                return dummy.next
            
            # Finding the midpoint of the list
            mid = find_mid(sta)
            next_to_mid = mid.next
            mid.next = None
            next_to_mid.prev = None
            
            # Recursively sort both halves
            left = mergesort(sta)
            right = mergesort(next_to_mid)
            
            # Merge the sorted halves
            return merge(left, right)
        
        return mergesort(head)

    #Your code End
    


#{ 
 # Driver Code Starts
#Initial Template for Python 3

import sys

sys.setrecursionlimit(1000000)


class Node:

    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class DoublyLinkedList:

    def __init__(self):
        self.head = None
        self.tail = None

    def append(self, new_data):
        new_node = Node(new_data)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
            return
        new_node.prev = self.tail
        self.tail.next = new_node
        self.tail = new_node

    def printList(self, node):
        while (node.next is not None):
            node = node.next
        while node.prev is not None:
            node = node.prev
        while (node is not None):
            print(node.data, end=" ")
            node = node.next
        print()


def printList(node):
    temp = node

    while (node is not None):
        print(node.data, end=" ")
        temp = node
        node = node.next
    print()
    while (temp):
        print(temp.data, end=" ")
        temp = temp.prev


if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        n = int(input())
        arr = list(map(int, input().strip().split()))
        llist = DoublyLinkedList()
        for e in arr:
            llist.append(e)
        ob = Solution()
        llist.head = ob.sortDoubly(llist.head)
        printList(llist.head)
        print()

# } Driver Code Ends
